java-checks/src/main/java/org/sonar/java/checks/DoubleCheckedLockingAssignmentCheck.java

```diff
@@ -0,0 +1,184 @@
+import org.sonar.plugins.java.api.tree.MemberSelectExpressionTree;
+import org.sonar.plugins.java.api.tree.StatementTree;
+import org.sonar.plugins.java.api.tree.SynchronizedStatementTree;
+import org.sonar.plugins.java.api.tree.Tree;
+
+@Rule(key = "S3064")
+public class DoubleCheckedLockingAssignmentCheck extends IssuableSubscriptionVisitor {
+
+  @Override
+  public List<Tree.Kind> nodesToVisit() {
+    return Collections.singletonList(Tree.Kind.IF_STATEMENT);
+  }
+
+  @Override
+  public void visitNode(Tree tree) {
+    if (!hasSemantic()) {
+      return;
+    }
+    IfStatementTree ifStatementTree = (IfStatementTree) tree;
+    StatementTree body = getOnlyStatement(ifStatementTree);
+    if (body == null || !body.is(Tree.Kind.SYNCHRONIZED_STATEMENT)) {
+      return;
+    }
+    List<StatementTree> synchronizedBody = ((SynchronizedStatementTree) body).block().body();
+    if (synchronizedBody.isEmpty()) {
+      return;
+    }
+    Symbol symbol = nullCheck(ifStatementTree.condition());
+    if (symbol == null || !symbol.owner().isTypeSymbol()) {
+      return;
+    }
+    for (StatementTree statementTree : synchronizedBody) {
+      if (statementTree.is(Tree.Kind.IF_STATEMENT)) {
+        IfStatementTree secondIf = (IfStatementTree) statementTree;
+        Symbol secondIfSymbol = nullCheck(secondIf.condition());
+        if (symbol.equals(secondIfSymbol)) {
+          checkUsageAfterAssignment(symbol, secondIf);
+        }
+      }
+    }
+  }
+
+  private void checkUsageAfterAssignment(Symbol symbol, IfStatementTree secondIf) {
+    if (secondIf.thenStatement().is(Tree.Kind.BLOCK)) {
+      List<StatementTree> body = ((BlockTree) secondIf.thenStatement()).body();
+      AssignmentExpressionTree foundAssignment = null;
+      UsageVisitor usageVisitor = new UsageVisitor(symbol);
+      for (StatementTree statementTree : body) {
+        if (foundAssignment != null) {
+          statementTree.accept(usageVisitor);
+        }
+        AssignmentExpressionTree assignment = isAssignmentToSymbol(statementTree, symbol);
+        if (assignment != null) {
+          foundAssignment = assignment;
+        }
+      }
+      if (foundAssignment != null && !usageVisitor.usages.isEmpty()) {
+        reportIssue(foundAssignment, "Fully initialize \"" + symbol.name() + "\" before assigning it.", usageVisitor.locations(), null);
+      }
+    }
+  }
+
+  @CheckForNull
+  private AssignmentExpressionTree isAssignmentToSymbol(StatementTree statementTree, Symbol symbol) {
+    if (!statementTree.is(Tree.Kind.EXPRESSION_STATEMENT)) {
+      return null;
+    }
+    ExpressionTree expr = ((ExpressionStatementTree) statementTree).expression();
+    if (expr.is(Tree.Kind.ASSIGNMENT)) {
+      AssignmentExpressionTree assignment = (AssignmentExpressionTree) expr;
+      if (symbol.equals(symbol(assignment.variable()))) {
+        return assignment;
+      }
+    }
+    return null;
+  }
+
+  static class UsageVisitor extends BaseTreeVisitor {
```

Comment:
I think you should skip lambdas and classes or methods. You may have inner classes, or anonymous classes using this symbol.